src/garrysmod/gamemodes/zombiesurvival/gamemode/sv_database.lua

```diff
@@ -2,31 +2,130 @@ require "pg"
+GM.Database._Connected = false
+GM.Database._Queue = {}
 
-function GM.Database:EstablishConnection(onConnected)
+function GM.Database:Connect()
     local host, user, password, port, database = host:GetString(), user:GetString(), password:GetString(), database:GetString()
-
+    
     if pg then
-        self.conn = pg.new_connection()
-
-        local success, err = self.conn:connect(host, user, password, database, 5432)
-
+        self.Connection = pg.new_connection()
+        
+        local success, err = self.Connection:connect(host, user, password, database, 5432)
+        
         if success then
-            success, err = self.conn:set_encoding("UTF8")
-
+            success, err = self.Connection:set_encoding("UTF8")
+            
             if not success then
                 ErrorNoHalt("Failed to set connection encoding:\n", err)
             end
-
-            if isfunction(onConnected) then onConnected(self) end
+            
+            MsgN("Connected to the database!")
+            
+            self._Connected = true
+            hook.Run("DatabaseConnected")
         else
             ErrorNoHalt("Unable to connect to the database!\n", err)
+            hook.Run("DatabaseConnectionFailed", error_text)
         end
     else
         ErrorNoHalt("gmsv_pg has not been found.")
     end
 end
+
+function GM.Database:Connected()
+    if self._Connected then
+        local _,is_open = pcall(function() return self.Connection:is_open() end)
+
+        -- If a connection has already been established, we can check the status
+        -- via pqxx::connection::is_open. Otherwise, use pcall to see if the "pg
+        -- - no connection, connect to a database first." error has been thrown.
+        if isbool(is_open) then
+            return is_open
+        else
+            return false
+        end
+    else
+        return false
+    end
+end
+
+function GM.Database:Escape(str)
+    return self.Connection:escape(str)
+end
+
+function GM.Database:Quote(str)
+    return self.Connection:quote(str)
+end
+
+function GM.Database:QuoteName(str)
+    return self.Connection:quote_name(str)
+end
+
+function GM.Database:Queue(query, callback)
+    assert(type(query) == "string")
+    table.insert(self._Queue, { query, callback })
+end
+
+function GM.Database:RawQuery(query, callback)
+    if not self:Connected() then
+        return self:Queue(query, callback)
+    end
+    
+    local query_obj = self.Connection:query(query)
+    query_obj:set_sync(true)
+    
+    local success, res, size = query_obj:run()
+    if success then
+        if isfunction(callback) then return callback(res, size) end
+    else
+        ErrorNoHalt("PostgreSQL Query Error!\n")
+        ErrorNoHalt(string.format("Query: %s\n", query))
+        local err = tostring(res)
+        ErrorNoHalt(string.format("%s\n", err))
+
+        if err == "Connection to database failed" then
+            return self:Queue(query, callback)
+        end
```

Comment:
When the database is dead, `pqxx::connection::is_open` will return `true` until the first [`pqxx::broken_connection`](https://libpqxx.readthedocs.io/en/latest/a00854.html) exception is thrown.

So, just for this first failed query, we need to punt it back into the queue here.